### DISCUSSION TITLE
Variable in x-for not working properly

### DISCUSSION THREAD

**Discussion by dennisvanderweide** [alpinejs/alpine#3614] [7. Bugs (archived)]
There is some weird behavior when using `x-for`.

Loop variable comparisons don't seem to work properly

I have the following code

``` JS
<template x-for="page in pages()">
    <div>
        <span x-text="page"></span>
        <span x-text="currentPage"></span>

        <div>
            <span x-text="page == currentPage"></span>
            <span x-text="page"></span>
        </div>
    </div>
</template>
```

This can sometimes result in the following UI
![image](https://github.com/alpinejs/alpine/assets/32065278/4b042d3a-b592-48f3-afa1-65eb0dc28be8)

As you can see `currentPage = 4` and `page = 3` but `page == currentPage` gives `true`
And where `currentPage = 4` and `page = 4` `page == currentPage` gives `false`

I am currently using `3.10.5` but I checked and the issue also exists in `3.12.2`

**Comment by ekwoka** [comment]
Can you show the data this is using?

x-for should also always have a `:key` which can do a lot to help.

when I test this same code it works fine:

<img width="83" alt="Screenshot 2023-06-07 at 4 09 12 PM" src="https://github.com/alpinejs/alpine/assets/43540491/f3910118-e149-479e-a006-a3228326211a">
